kad-dht/README.md

```diff
@@ -0,0 +1,395 @@
+
+	// Validate validates the given record, returning an error if it's
+	// invalid (e.g., expired, signed by the wrong key, etc.).
+	Validate(key string, value []byte) error
+
+	// Select selects the best record from the set of records (e.g., the
+	// newest).
+	//
+	// Decisions made by select should be stable.
+	Select(key string, values [][]byte) (int, error)
+}
+```
+
+`Validate()` is a pure function that reports the validity of a record. It may
+validate a cryptographic signature, or else. It is called on two occasions:
+
+1. To validate incoming values in response to `GET_VALUE` calls.
+2. To validate outgoing values before storing them in the network via
+   `PUT_VALUE` calls.
+
+Similarly, `Select()` is a pure function that returns the best record out of 2
+or more candidates. It may use a sequence number, a timestamp, or other
+heuristic to make the decision.
+
+## Public key records
+
+Apart from storing arbitrary values, the libp2p Kad DHT stores node public keys
+in records under the `/pk` namespace. That is, the entry `/pk/<peerID>` will
+store the public key of peer `peerID`.
+
+DHT implementations may optimise public key lookups by providing a
+`GetPublicKey(peer.ID) (ci.PubKey)` method, that, for example, first checks if
+the key exists in the local peerstore.
+
+The lookup for public key entries is identical to a standard entry lookup,
+except that a custom `Validator` strategy is applied. It checks that equality
+`SHA256(value) == peerID` stands when:
+
+1. Receiving a response from a `GET_VALUE` lookup.
+2. Storing a public key in the DHT via `PUT_VALUE`.
+
+The record is rejected if the validation fails.
+
+## Provider records
+
+Nodes must keep track of which nodes advertise that they provide a given key
+(CID). These provider advertisements should expire, by default, after 24 hours.
+These records are managed through the `ADD_PROVIDER` and `GET_PROVIDERS`
+messages.
+
+When `Provide(key)` is called, the DHT finds the closest peers to `key` using
+the `FIND_NODE` RPC, and then sends a `ADD_PROVIDER` RPC with its own
+`PeerInfo` to each of these peers.
+
+Each peer that receives the `ADD_PROVIDER` RPC should validate that the
+received `PeerInfo` matches the sender's `peerID`, and if it does, that peer
+must store a record in its datastore the received `PeerInfo` record.
+
+When a node receives a `GET_PROVIDERS` RPC, it must look up the requested
+key in its datastore, and respond with any corresponding records in its
+datastore, plus a list of closer peers in its routing table.
+
+For performance reasons, a node may prune expired advertisements only
+periodically, e.g. every hour.
+
+## Node lookups
+
+_WIP (raulk)._
+
+## Bootstrap process
+The bootstrap process is responsible for keeping the routing table filled and
+healthy throughout time. It runs once on startup, then periodically with a
+configurable frequency (default: 5 minutes).
+
+On every run, we generate a random node ID and we look it up via the process
```

Comment:
>we generate a random node ID

Specify how node ID is generated.
Is it the same as peer ID?

Reply:
Simplified by using `peer ID` with https://github.com/libp2p/specs/pull/108/commits/a9ec52376b0cd923f8f1c6a8e6919c10d763805f.